Created developer documentation (#90)
Redoublet-org/Redoublet-website: README.md

Reviewer: Ik mis nog een sectie over de dev-omgeving

# Redoublet-website
This repository contains the code for the Vue.js frontend of the site in the Redoublet-frontend folder and the C# game logic backend in the Redoublet-backend folder.

[Firebase](https://console.firebase.google.com/u/0/project/redoublet-site/overview) is used for the authentication and for keeping the state of each multiplayer game, where as Robobridge is used for generating moves for the robot players.

![Architecture](documentation-assets\Architecture_1.png)

## How to run the code locally?
Prequisites:

> - Docker Desktop

Reviewer: ```suggestion
- [Docker Desktop](https://www.docker.com/products/docker-desktop/)
```

Steps in order to run the frontend and game logic backend locally:
1. Clone this repo
2. Navigate to the repo location
3. Open a command prompt and run the following commands (make sure Docker Desktop is running):
```

> docker-compose build

Reviewer: ```suggestion
(docker-compose down)
docker-compose build
```

docker-compose up
```

> We have created a cmd shortcut for this.

Reviewer: ```suggestion
We have created a Command Prompt and a Powershell shortcut for this which runs the 3 commands in the following order and then also starts the website.
```

> In CMD:

Reviewer: ```suggestion

In Command Prompt:
```

```
up
```

In Powershell:
```
.\up
```

## Multiplayer gamestate synchronisation

> Because the API inside this repo doesn't keep track of the gamestate we use [Firebase Realtime Database](https://console.firebase.google.com/u/0/project/redoublet-site/database/redoublet-site-default-rtdb/data). Here we keep track of every multiplayer game that is played on the site. The content of the Realtime Database is basically a JSON tree containing a list of games with the info required to play a game:

Reviewer: ```suggestion
Because the API inside this repository doesn't keep track of the gamestate we use [Firebase Realtime Database](https://console.firebase.google.com/u/0/project/redoublet-site/database/redoublet-site-default-rtdb/data). Here we keep track of every multiplayer game that is played on the site. The content of the Realtime Database is basically a JSON tree containing a list of games with the info required to play a game. Here is an example of what it looks like:
```

  - currentPlayerID : "12345"           (ID of the player whose turn it is)
  - host : "John Doe"                   (username of the player who hosted the game)
  - state : "started"                   (phase of the game: waiting, started, finished)
  - players                             (list containing the players in the game)
    - East : "12345"                        (the position at the table of each player)
    - North : "54321"
    - South : "98765"
    - West : "67890" 
  - gamestate                           (parent object of game specifics such as the cards)
```

In Vue.js we upload the new gamestate to the Realtime Firebase we received from the game logic API after playing a move. We have defined Realtime Database rules to ensure only the player whose turn it is can edit the gamestate in the database.

## Authentication
We use [Firebase Authentication](https://console.firebase.google.com/u/0/project/redoublet-site/authentication/users) to allow users to register and login with an account. Official documentation for Firebase Authentication relevant to the Vue.js frontend can be found [here](https://firebase.google.com/docs/auth/web/start).

## CI/CD

> Every push to the production branch in this repo will [trigger](https://console.cloud.google.com/cloud-build/triggers?project=redoublet-site) dockerize and build th back- and frontend in Google Cloud Build and deploy the backend API to [Google Cloud Run](https://console.cloud.google.com/run/detail/europe-west4/backend/metrics?project=redoublet-site) and the frontend to [Firebase Hosting](https://console.firebase.google.com/u/0/project/redoublet-site/hosting/sites).

Reviewer: ```suggestion
Every push to the production branch in this repository will [trigger](https://console.cloud.google.com/cloud-build/triggers?project=redoublet-site) dockerize and build the backend and frontend in Google Cloud Build and deploy the backend API to [Google Cloud Run](https://console.cloud.google.com/run/detail/europe-west4/backend/metrics?project=redoublet-site) and the frontend to [Firebase Hosting](https://console.firebase.google.com/u/0/project/redoublet-site/hosting/sites).
```

Voeg hier ook nog dat stappenplan toe wat ik op discord heb gegeven mbt. het aanmaken van nieuwe triggers.

# Redoublet-website
This repository contains the code for the Vue.js frontend of the site in the Redoublet-frontend folder and the C# game logic backend in the Redoublet-backend folder.

[Firebase](https://console.firebase.google.com/u/0/project/redoublet-site/overview) is used for the authentication and for keeping the state of each multiplayer game, where as [RoboBridge](https://www.robobridge.com) is used for generating moves for the robot players.

![Architecture](documentation-assets/Architecture_1.png)

## How to run the code locally
Prequisites:
- [Docker Desktop](https://www.docker.com/products/docker-desktop/)

Steps in order to run the frontend and game logic backend locally:
1. Clone this repository
2. Navigate to the repository location
3. Open a command prompt and run the following commands (make sure Docker Desktop is running).

optional:
```
docker-compose down
```
then:
```
docker-compose build
docker-compose up
```

We have created a Command Prompt and a Powershell shortcut for this which runs the 3 commands in the following order and then also starts the website.

> In Command Prompt:

Reviewer: ```suggestion

In Command Prompt:
```

We use [Firebase Authentication](https://console.firebase.google.com/u/0/project/redoublet-site/authentication/users) to allow users to register and login with an account. Official documentation for Firebase Authentication relevant to the Vue.js frontend can be found [here](https://firebase.google.com/docs/auth/web/start).

## CI/CD

> We have two different branches, the production branch is always live on [redoublet.com](https://redoublet.com) and the main branch is live on [dev.redoublet.com](https://dev.redoublet.com).

Reviewer: ```suggestion
We have two different branches, the production `prod` branch is always live on [redoublet.com](https://redoublet.com) and the developer `main` branch is always live on [dev.redoublet.com](https://dev.redoublet.com).
```

Every push to the production branch in this repository will trigger the [prod-trigger](https://console.cloud.google.com/cloud-build/triggers?project=redoublet-site) and dockerize + build the backend in Google Cloud Build and deploy the backend API to [Google Cloud Run](https://console.cloud.google.com/run/detail/europe-west4/backend/metrics?project=redoublet-site) and deploy the frontend to [Firebase Hosting](https://console.firebase.google.com/u/0/project/redoublet-site/hosting/sites).

> The same thing happens for pushes to the main branch, this will trigger the [main-trigger](https://console.cloud.google.com/cloud-build/triggers?project=redoublet-site-dev). This can be used as a development environment. You can follow these steps in order to get another branch deployed to [dev.redoublet.com](https://dev.redoublet.com):

Reviewer: ```suggestion
The same thing happens for pushes to the developer branch. This will trigger the [main-trigger](https://console.cloud.google.com/cloud-build/triggers?project=redoublet-site-dev). This can be used as a development environment. You can follow these steps in order to get another branch deployed to [dev.redoublet.com](https://dev.redoublet.com):
```
